### ISSUE TITLE
RasterImage resizeTo() and copy() don't do anything

### ISSUE THREAD

**Issue by marcusnaslund** [magic-lang/ooc-kean#574]
The old comments `// TODO: Actually resize the image` and my personal favourite `// TODO: The stuff` haven't been taken care of yet.

We obviously don't use these (or we would have fixed it much sooner) but we also don't implement these functions in all derived classes. The function **is** there, though, but doesn't do what it should. Hence the bug label.

**Comment by davidhesselbom** [comment]
Actually, we may yet have use for these functions &mdash; I'm not sure exactly what's needed by the CPU version of the project. Hold this off until after that deadline, or we'll likely just end up with conflicts where someone implemented the function for real and someone else deleted it.

**Comment by marcusnaslund** [comment]
Oh, I wasn't saying delete. Better to fix them if there's a chance we might need them.

**Comment by sebastianbaginski** [comment]
Please don't delete resizeTo: https://github.com/cogneco/ooc-kean/pull/530

**Comment by sebastianbaginski** [comment]
So, `copy()` that takes a `transform` as parameter should behave something like `cvWarpPerspective` (http://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html#warpperspective) ? I will be implementing this, unless someone has objections.

**Comment by sebastianbaginski** [comment]
It basically works but I could use some help on the interface. Now there is a `copy` method that takes `IntSize2D` and `FloatTransform2D`. I don't think this is a clear interface, as I don't know what it is supposed to do just by looking at it. Does `copy(size, transform)` mean "resize this image to given size and then transform it" or something else ? If this is what it's supposed to do, then the `size` part is redundant as rescaling information can be included in the `transform` parameter.
What about creating a method `transformed: func (transform: FloatTransform2D) -> This` ? I think it is clear what this function will do ("return a copy of this image transformed by given transform").
Furthermore, how do we deal with GPU part ? I know that there is a "transform" parameter already in the "GpuCanvas". @emilwestergren do you have any suggestions about how do we unify the interface ? Should the transform be a part of `AbstractCanvas` instead ? So the interface to get transformed image will be
```ooc
transform := FloatTransform2D createScaling(2,2)
image canvas transform = transform
result := image somefunction() // I don't know how we should call it in the "canvas" case
```
instead of
```ooc
transform := FloatTransform2D createScaling(2,2)
result := image transformed(transform)
```
I like the second better, but I suppose we should aim to make it identical for GPU and CPU images.
Help appreciated.

**Comment by emilwestergren** [comment]
Transform is already defined in the abstract Canvas class. I agree that the current Image API is very confusing and could probably be improved in some way. First of all there is a distinction between the two different APIs:

1. The simple Image API with primitive functions (`image copy()` `image resizeTo()` etc)
2. The more powerful Draw API (`image canvas draw(points)` `image canvas draw(map)` etc)

The Image API often makes use of the Draw API, but provides a simpler interface for primitive operations. So if you are changing the Image API then I would leave the Draw API as it is for now and deal with it separately.

**Comment by sebastianbaginski** [comment]
OK, thanks, that makes sense. Any idea about what `copy(size, transform)` is supposed to do ?
I'd just get rid of it if you ask me.

**Comment by emilwestergren** [comment]
I think it's just a simplification (and optimization) of transforming it and rescaling it. You can achieve the same thing with a single transform but then you would have to calculate the scaling factor by yourself. I don't think it's used anywhere in the code so I doubt anyone would miss it.

**Comment by marcusnaslund** [comment]
@sebastianbaginski - this is fixed, right?

**Comment by sebastianbaginski** [comment]
Not yet, I will be working on this soon.

**Comment by marcusnaslund** [comment]
Fixed by #1706
